File: backend/internal/api/middleware.go
package api

import (
	"log/slog"
	"net/http"
	"net/url"
	"runtime/debug"
	"time"
)

// isLocalOrigin reports whether an Origin header points at a loopback host,
// which is the only cross-origin caller this local game server expects.
func isLocalOrigin(origin string) bool {
	u, err := url.Parse(origin)
	if err != nil {
		return false
	}
	host := u.Hostname()
	return u.Scheme == "http" && (host == "localhost" || host == "127.0.0.1" || host == "::1")
}

func CORSMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		origin := r.Header.Get("Origin")
		if origin != "" && isLocalOrigin(origin) {
			w.Header().Set("Access-Control-Allow-Origin", origin)
			w.Header().Set("Access-Control-Allow-Credentials", "true")
			w.Header().Set("Access-Control-Allow-Methods", "GET, POST, DELETE, OPTIONS")
			w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
		}
		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusNoContent)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func LoggingMiddleware(logger *slog.Logger, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		logger.Info("request", "method", r.Method, "path", r.URL.Path, "duration", time.Since(start))
	})
}

func RecoveryMiddleware(logger *slog.Logger, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if rec := recover(); rec != nil {
				logger.Error("panic recovered", "error", rec, "stack", string(debug.Stack()))
				writeJSON(w, http.StatusInternalServerError, ErrorResponse{Error: "internal", Message: "Internal server error"})
			}
		}()
		next.ServeHTTP(w, r)
	})
}

File: backend/src/Caro.Api/Middleware.cs
using Caro.Domain;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;

namespace Caro.Api;

public static class LocalOrigin
{
    private const string LoopbackScheme = "http";

    // Uri.Host keeps the brackets on IPv6 ("[::1]"); Go's Hostname()
    // stripped them, so both spellings are accepted.
    private static readonly string[] LoopbackHosts = ["localhost", "127.0.0.1", "::1", "[::1]"];

    /// <summary>
    /// Reports whether an Origin header points at a loopback host, which is
    /// the only cross-origin caller this local game server expects.
    /// </summary>
    public static bool IsLocalOrigin(string origin)
    {
        if (!Uri.TryCreate(origin, UriKind.Absolute, out Uri? u))
        {
            return false;
        }
        return u.Scheme == LoopbackScheme && LoopbackHosts.Contains(u.Host);
    }
}

/// <summary>
/// Maps domain exceptions to the HTTP error contract and converts anything
/// unexpected into a 500, mirroring the Go writeError/panic-recovery pair.
/// </summary>
public sealed class ErrorMappingMiddleware(RequestDelegate next, ILogger<ErrorMappingMiddleware> logger)
{
    public async Task InvokeAsync(HttpContext http)
    {
        try
        {
            await next(http);
        }
        catch (Exception e)
        {
            switch (e)
            {
                case GameNotFoundException:
                    await ResponseJson.Write(http, 404, new ErrorResponse { Error = "not_found", Message = e.Message });
                    break;
                case TooManyGamesException:
                    await ResponseJson.Write(http, 429, new ErrorResponse { Error = "too_many_games", Message = e.Message });
                    break;
                case CellOccupiedException or PositionBoundsException or GameOverException
                    or OpenRuleException or InvalidLevelException:
                    await ResponseJson.Write(http, 400, new ErrorResponse { Error = "bad_request", Message = e.Message });
                    break;
                case NotPlayerTurnException:
                    await ResponseJson.Write(http, 409, new ErrorResponse { Error = "not_your_turn", Message = e.Message });
                    break;
                default:
                    logger.Panic(e, http.Request.Path.ToString());
                    await ResponseJson.Write(http, 500, new ErrorResponse { Error = "internal", Message = "Internal server error" });
                    break;
            }
        }
    }
}

public sealed class RequestLoggingMiddleware(RequestDelegate next, ILogger<RequestLoggingMiddleware> logger)
{
    public async Task InvokeAsync(HttpContext http)
    {
        long start = System.Diagnostics.Stopwatch.GetTimestamp();
        await next(http);
        if (logger.IsEnabled(LogLevel.Information))
        {
            double durationMs = System.Diagnostics.Stopwatch.GetElapsedTime(start).TotalMilliseconds;
            logger.Request(http.Request.Method, http.Request.Path.ToString(), durationMs);
        }
    }
}
